core/domain/classifier_services.py

```diff
@@ -0,0 +1,322 @@
+        w - word
+        d - doc
+
+        Internal (private) model representation:
+        _w_dc - word (id) given doc and count
+            (list of word ids which represent a doc)
+        _b_dl - boolean given doc and label
+            (boolean of whether a doc has a label)
+        _l_dc - label given doc and count
+            (assigned label of each word in a doc)
+        _c_dl - count given doc and label
+            (count of each label per doc)
+        _c_lw - count given label and word
+            (count of labels per word)
+        _c_l - count given label (count of each label)
+        """
+        # Ensure that results are deterministic
+        # There is nothing special about the value 4
+        numpy.random.seed(seed=4)
+
+        self._alpha = self._DEFAULT_ALPHA
+        self._beta = self._DEFAULT_BETA
+
+    def _get_word_id(self, word):
+        """Returns a word's id if it exists, otherwise assigns
+        a new id to the word and returns it."""
+        try:
+            return self._word_to_id[word]
+        except KeyError as ke:
+            self._word_to_id[word] = self._word_count
+            self._word_count += 1
+        return self._word_to_id[word]
+
+    def _get_label_id(self, label):
+        """Returns a word's id if it exists, otherwise assigns
+        a new id to the label and returns it."""
+        try:
+            return self._label_to_id[label]
+        except KeyError as ke:
+            self._label_to_id[label] = self._label_count
+            self._label_count += 1
+        return self._label_to_id[label]
+
+    def _get_record(self, d):
+        """Given a doc id, return the doc and labels."""
+        return self._w_dc[d], self._b_dl[d]
+
+    def _get_label_vector(self, labels):
+        """Returns a vector specifying which labels
+        are specified in the training set."""
+        if len(labels) == 0:
+            return numpy.ones(self._label_count)
+        label_vector = numpy.zeros(self._label_count)
+        for label in labels:
+            label_vector[self._get_label_id(label)] = 1
+        # Always set default label
+        label_vector[self._label_to_id['_default']] = 1
+        return label_vector
+
+    def _update_counting_matrices(self, d, w, l, val):
+        """Update counting matrices given doc/word/label indices."""
+        self._c_dl[d, l] += val
+        self._c_lw[l, w] += val
+        self._c_l[l] += val
+
+    def _get_doc_ids(self, doc_ids):
+        if doc_ids is None:
+            doc_ids = xrange(self._doc_count)
+        return doc_ids
+
+    def _init_docs(self, doc_ids=None):
+        """Initialize data for given docs (defaults to all)."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for d in doc_ids:
+            doc, labels = self._get_record(d)
+            l_c = [
+                numpy.random.multinomial(1, labels / labels.sum()).argmax()
+                for i in xrange(len(doc))
+            ]
+            self._l_dc.append(l_c)
```

Comment:
For the general case, it is the former. The assumption is that when a doc is added, we init it with some non-zero values once and only once. That is being enforced by functions that call _init_docs (unless people decide to break the rules and call it outside of StringClassifier).